astropy/cosmology/funcs/optimize.py

```diff
@@ -1,31 +1,104 @@
 # Licensed under a 3-clause BSD style license - see LICENSE.rst
 """Convenience functions for `astropy.cosmology`."""
 
+from __future__ import annotations
+
 import warnings
+from typing import TYPE_CHECKING, Any, NotRequired, Protocol, TypedDict
 
 import numpy as np
+import numpy.typing as npt
 
 from astropy.cosmology import units as cu
 from astropy.cosmology.core import CosmologyError
 from astropy.units import Quantity
 from astropy.utils.exceptions import AstropyUserWarning
 
+if TYPE_CHECKING:
+    from collections.abc import Callable
+    from typing import TypeAlias
+
+    from scipy.optimize import OptimizeResult
+
 __all__ = ["z_at_value"]
 
 __doctest_requires__ = {"*": ["scipy"]}
 
 
+# TODO: it would be nice to upstream this to scipy and then use it here.
+class _CustomMethod(Protocol):
```

Comment:
Why is this implemented as a `Protocol` with a `__call__()` method instead of something like
```python
_CustomMethod: TypeAlias = Callable[..., OptimizeResult]
```
And is `_CustomMethod` the best name for this?

Reply:
Because the structure of `_CustomMethod` is most specific than that. It has to have a signature like 
`(fun: Callable[..., Any], args: tuple[Any, ...], **kwargs: Any) -> OptimizeResult`.
But I agree about a better name!